### PR TITLE
refactor(worker): refact plugin & add script plugin

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18726.2 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18748.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18748.1 ■   
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ■   
  * Standalone Tests ■

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18748.1 ■   
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ■   
  * Standalone Tests ■

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18748.2 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18750.0 ■   
* Build
  * Build and Package All ■   
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18750.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18750.2 ■   
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ■   
  * Standalone Tests ■

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18750.2 ■   
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ■   
  * Standalone Tests ■

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18750.3 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18751.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+			gores.Status = sdk.StatusFail
+			gores.Details = fmt.Sprintf("%v", err)
+			chanRes <- gores
+			return
+		}
+
+		cmd := exec.CommandContext(ctx, script.shell, script.opts...)
+		pr, pw := io.Pipe()
+		cmd.Dir = script.dir
+		//cmd.Env = wk.Environ()
```
Comment (yesnault) [inline contrib/grpcplugins/action/plugin-script/main.go:94]:
```suggestion
		//cmd.Env = wk.Environ()
```

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+		_ = pr.Close()
+		<-outchan
+
+		chanRes <- gores
+	}()
+
+	res := &actionplugin.ActionResult{}
+	select {
+	case <-ctx.Done():
+		fmt.Printf("\"CDS Worker execution canceled: %v", ctx.Err())
```
Comment (yesnault) [inline contrib/grpcplugins/action/plugin-script/main.go:151]:
```suggestion
		fmt.Printf("CDS Worker execution canceled: %v", ctx.Err())
```

File: engine/worker/internal/run_ascode.go [hunk 3]
```diff
@@ -23,46 +22,70 @@ func (w *CurrentWorker) runAscodeAction(ctx context.Context, actionName string)
 	for _, step := range currentAscodeAction.Runs.Steps {
 		switch {
 		case step.Uses != "":
-			log.Debug(ctx, "running subaction %s", step.Uses)
+			actionName := strings.TrimPrefix(step.Uses, "actions/")
+			actionPath := strings.Split(actionName, "/")
+			stepName := prefixStepName + filepath.Base(actionName)
```
Comment (yesnault) [inline engine/worker/internal/run_ascode.go:27]: to recheck, probably missing '/'

File: engine/worker/internal/run_ascode.go [hunk 4]
```diff
@@ -23,46 +22,70 @@ func (w *CurrentWorker) runAscodeAction(ctx context.Context, actionName string)
...
+func (wk *CurrentWorker) runPlugin(ctx context.Context, pluginName string, opts map[string]string) sdk.Result {
+	pluginClient, err := plugin.NewClient(ctx, wk, plugin.TypeAction, pluginName)
+	if pluginClient != nil {
+		defer pluginClient.Close(ctx)
+	}
+	if err != nil {
+		return wk.failAction(ctx, fmt.Sprintf("%v", err))
+	}
+
+	// TODO Set input regarding plugin descriptor + interpoloate action params with context
```
Comment (yesnault) [inline engine/worker/internal/run_ascode.go:85]: TODO ?

**Review by yesnault — CHANGES_REQUESTED** [review]

File: engine/worker/internal/run_ascode.go (hunk 3 above)
Comment (sguiheux) [reply engine/worker/internal/run_ascode.go:27]: prefix contains /

File: contrib/grpcplugins/action/plugin-archive/go.mod [hunk 5]
```diff
@@ -68,6 +68,7 @@ require (
 	github.com/sirupsen/logrus v1.9.0 // indirect
 	github.com/spf13/afero v1.8.1 // indirect
 	github.com/spf13/cast v1.4.1 // indirect
+	github.com/tebeka/go2xunit v1.4.10 // indirect
```
Comment (richardlt) [inline contrib/grpcplugins/action/plugin-archive/go.mod:71]: To remove

Comment (richardlt) [inline contrib/grpcplugins/action/plugin-archive/go.mod:71]: To remove

File: contrib/grpcplugins/action/plugin-archive/go.sum [hunk 6]
```diff
@@ -322,6 +322,8 @@ github.com/stretchr/testify v1.7.2/go.mod h1:R6va5+xMeoiuVRoj+gSkQ7d3FALtqAAGI1F
 github.com/stretchr/testify v1.8.0 h1:pSgiaMZlXftHpm5L7V1+rVB+AZJydKsMxsQBIJw4PKk=
 github.com/stretchr/testify v1.8.0/go.mod h1:yNjHg4UonilssWZ8iaSj1OCr/vHnekPRkoO+kdMU+MU=
 github.com/subosito/gotenv v1.2.0 h1:Slr1R9HxAlEKefgq5jn9U+DnETlIUa6HfgEzj0g5d7s=
+github.com/tebeka/go2xunit v1.4.10 h1:0UO+9YoLpXTZ0DL9XbTmIIibgmKBGiwroo8uhFMSyR0=
```
Comment (richardlt) [inline contrib/grpcplugins/action/plugin-archive/go.sum:325]: To remove

**Review by richardlt — APPROVED** [review]

File: engine/worker/internal/plugin/plugin.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+	"github.com/ovh/cds/sdk/grpcplugin/actionplugin"
+	"github.com/ovh/cds/sdk/grpcplugin/integrationplugin"
+)
+
+// NewClient create a plugin client for the given plugin
+func NewClient(ctx context.Context, wk workerruntime.Runtime, pluginType string, pluginName string) (Client, error) {
+	// Create socket
+	pluginSocket, currentPlugin, err := createGRPCPluginSocket(ctx, pluginType, pluginName, wk)
+	if err != nil {
+		return nil, fmt.Errorf("unable to start GRPCPlugin: %v", err)
```
Comment (fsamin) [inline engine/worker/internal/plugin/plugin.go:24]:
```suggestion
		return nil, errors.Errorf("unable to start GRPCPlugin: %v", err)
```

File: engine/worker/internal/plugin/plugin.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+		pluginType: pluginType,
+		grpcPlugin: currentPlugin,
+	}
+
+	switch pluginType {
+	case TypeAction:
+		// Create grpc client
+		grpcClient, err := actionplugin.Client(context.Background(), pluginSocket.Socket)
+		if err != nil {
+			return nil, fmt.Errorf("unable to call GRPCPlugin: %v", err)
```
Comment (fsamin) [inline engine/worker/internal/plugin/plugin.go:41]:
```suggestion
			return nil, errors.Errorf("unable to call GRPCPlugin: %v", err)
```

File: engine/worker/internal/plugin/plugin.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+		}
+		qPort := actionplugin.WorkerHTTPPortQuery{Port: wk.HTTPPort()}
+		if _, err := grpcClient.WorkerHTTPPort(ctx, &qPort); err != nil {
+			return nil, fmt.Errorf("unable to setup plugin with worker port: %v", err)
```
Comment (fsamin) [inline engine/worker/internal/plugin/plugin.go:45]:
```suggestion
			return nil, errors.Errorf("unable to setup plugin with worker port: %v", err)
```

File: engine/worker/internal/plugin/plugin.go [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+		}
+		c.grpcClient = grpcClient
+	case TypeIntegration:
+		// Create grpc client
+		grpcClient, err := integrationplugin.Client(context.Background(), pluginSocket.Socket)
+		if err != nil {
+			return nil, fmt.Errorf("unable to call GRPCPlugin: %v", err)
```
Comment (fsamin) [inline engine/worker/internal/plugin/plugin.go:52]:
```suggestion
			return nil, errors.Errorf("unable to call GRPCPlugin: %v", err)
```

File: engine/worker/internal/plugin/plugin.go [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+		}
+		qPort := integrationplugin.WorkerHTTPPortQuery{Port: wk.HTTPPort()}
+		if _, err := grpcClient.WorkerHTTPPort(ctx, &qPort); err != nil {
+			return nil, fmt.Errorf("unable to setup plugin with worker port: %v", err)
```
Comment (fsamin) [inline engine/worker/internal/plugin/plugin.go:56]:
```suggestion
			return nil, errors.Errorf("unable to setup plugin with worker port: %v", err)
```

File: engine/worker/internal/plugin/plugin.go [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+	// Start plugin logger
+	sdk.NewGoRoutines(ctx).Run(ctx, "runGRPCPlugin", func(ctx context.Context) {
+		c.enablePluginLogger(logCtx)
+	})
+
+	// Test plugin
+	if err := c.Manifest(ctx); err != nil {
+		c.Close(ctx)
+		return nil, fmt.Errorf("unable to retrieve retrieve plugin manifest: %v", err)
```
Comment (fsamin) [inline engine/worker/internal/plugin/plugin.go:72]:
```suggestion
		return nil, errors.Errorf("unable to retrieve retrieve plugin manifest: %v", err)
```

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+	content := q.GetOptions()["content"]
+
+	workDirs, err := grpcplugins.GetWorkerDirectories(actPlugin.HTTPPort)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get working directory: %v", err)
+	}
+
+	chanRes := make(chan *actionplugin.ActionResult)
+
+	go func() {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:68]: done

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+			if strings.HasPrefix(cmd.Env[i], "PATH") {
+				cmd.Env[i] = fmt.Sprintf("%s:%s", cmd.Env[i], path.Dir(workerpath))
+				break
+			}
+		}
+
+		reader := bufio.NewReader(pr)
+
+		outchan := make(chan bool)
+		go func() {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:115]: done

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+	workDir, err := fs.Open(script.dir)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open working directory %s [%s]: %v", script.dir, filepath.Base(script.dir), err)
+	}
+	fi, err := workDir.Stat()
+	if err != nil {
+		return nil, err
+	}
+	if !fi.IsDir() {
+		return nil, fmt.Errorf("working directory %s is not a directory: %v", script.dir, err)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:207]: done

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	// Generate a random string 16 chars length
+	bs := make([]byte, 16)
+	if _, err := rand.Read(bs); err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:213]: done

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+
+	if isWindows() {
+		tmpFileName += ".PS1"
+	}
+
+	scriptPath := filepath.Join(path.Dir(script.dir), tmpFileName)
+
+	tmpscript, err := fs.OpenFile(scriptPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0700)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create temporary file, aborting: %v", err)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:225]: done

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	// Put script in file
+	n, err := tmpscript.Write(script.content)
+	if err != nil || n != len(script.content) {
+		if err != nil {
+			return nil, fmt.Errorf("unable to create script: %v", err)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:232]: done

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+		} else {
+			return nil, fmt.Errorf("unable to write all script: %d/%d", n, len(script.content))
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:234]: done

File: contrib/grpcplugins/action/plugin-script/main.go [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+		return nil, fmt.Errorf("unable to write script to %s", tmpscript)
+	}
+
+	var realScriptPath = scriptPath
+
+	switch x := fs.(type) {
+	case *afero.BasePathFs:
+		realScriptPath, err = x.RealPath(tmpscript.Name())
+		if err != nil {
+			return nil, fmt.Errorf("unable to get script working dir: %v", err)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/action/plugin-script/main.go:248]: done

File: contrib/grpcplugins/grpcplugins.go [hunk 21]
```diff
@@ -68,3 +68,34 @@ func SendVulnerabilityReport(workerHTTPPort int32, report sdk.VulnerabilityWorke
 
 	return nil
 }
+
+func GetWorkerDirectories(workerHTTPPort int32) (*sdk.WorkerDirectories, error) {
+	if workerHTTPPort == 0 {
+		return nil, fmt.Errorf("worker port must not be 0")
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/grpcplugins.go:74]: done

File: contrib/grpcplugins/grpcplugins.go [hunk 22]
```diff
@@ -68,3 +68,34 @@ func SendVulnerabilityReport(workerHTTPPort int32, report sdk.VulnerabilityWorke
...
+	}
+
+	req, err := http.NewRequest("GET", fmt.Sprintf("http://127.0.0.1:%d/directories", workerHTTPPort), nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create request to get directories: %v", err)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/grpcplugins.go:79]: done

File: contrib/grpcplugins/grpcplugins.go [hunk 23]
```diff
@@ -68,3 +68,34 @@ func SendVulnerabilityReport(workerHTTPPort int32, report sdk.VulnerabilityWorke
...
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get run result directories: %v", err)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/grpcplugins.go:84]: done

File: contrib/grpcplugins/grpcplugins.go [hunk 24]
```diff
@@ -68,3 +68,34 @@ func SendVulnerabilityReport(workerHTTPPort int32, report sdk.VulnerabilityWorke
...
+	}
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read body on get run result /working-directory: %v", err)
+	}
+
+	if resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("cannot get working directory: HTTP %d", resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/grpcplugins.go:93]: done

File: contrib/grpcplugins/grpcplugins.go [hunk 25]
```diff
@@ -68,3 +68,34 @@ func SendVulnerabilityReport(workerHTTPPort int32, report sdk.VulnerabilityWorke
...
+	}
+
+	var workDir sdk.WorkerDirectories
+	if err := sdk.JSONUnmarshal(body, &workDir); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal response: %v", err)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply contrib/grpcplugins/grpcplugins.go:98]: done

File: engine/worker/internal/plugin/plugin.go (hunk 7 above)
Comment (sguiheux) [reply engine/worker/internal/plugin/plugin.go:24]: done

File: engine/worker/internal/plugin/plugin.go (hunk 8 above)
Comment (sguiheux) [reply engine/worker/internal/plugin/plugin.go:41]: done

File: engine/worker/internal/plugin/plugin.go (hunk 9 above)
Comment (sguiheux) [reply engine/worker/internal/plugin/plugin.go:45]: done

File: engine/worker/internal/plugin/plugin.go (hunk 11 above)
Comment (sguiheux) [reply engine/worker/internal/plugin/plugin.go:56]: done

File: engine/worker/internal/plugin/plugin.go (hunk 12 above)
Comment (sguiheux) [reply engine/worker/internal/plugin/plugin.go:72]: done

**Review by fsamin — APPROVED** [review]

**Pull request merged**